pkg/azuredx/cache_test.go

```diff
@@ -0,0 +1,180 @@
+package azuredx
+
+import (
+	"testing"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewCacheSettings(t *testing.T) {
+	tests := []struct {
+		name          string
+		configuration dataSourceData
+		query         string
+		timeRange     backend.TimeRange
+		returnIs      assert.ComparisonAssertionFunc
+		returnVal     CacheSettings
+	}{
+		{
+			name: "should adjust settings based on query bin size 1m",
+			configuration: dataSourceData{
+				CacheMaxAge:    "10s",
+				DynamicCaching: true,
+			},
+			query: `SampleData\n` +
+				`| where state == "Texas"\n` +
+				`summarize avg(Coulumn) by bin(TimeColumn, 1m)`,
+
+			timeRange: backend.TimeRange{
+				From: time.Date(2019, 7, 30, 20, 2, 33, 0, time.UTC),
+				To:   time.Date(2019, 7, 30, 21, 7, 33, 0, time.UTC),
+			},
+			returnIs: assert.Equal,
+			returnVal: CacheSettings{
+				CacheMaxAge: "60s",
+				TimeRange: &backend.TimeRange{
+					From: time.Date(2019, 7, 30, 20, 2, 0, 0, time.UTC),
+					To:   time.Date(2019, 7, 30, 21, 8, 0, 0, time.UTC),
+				},
+			},
+		},
+		{
+			name: "should adjust settings based on query bin size 10m",
+			configuration: dataSourceData{
+				CacheMaxAge:    "10s",
+				DynamicCaching: true,
+			},
+			query: `SampleData\n` +
+				`| where state == "Texas"\n` +
+				`summarize avg(Coulumn) by bin(TimeColumn, 10m)`,
+			timeRange: backend.TimeRange{
+				From: time.Date(2019, 7, 30, 20, 2, 33, 0, time.UTC),
+				To:   time.Date(2019, 7, 30, 21, 7, 33, 0, time.UTC),
+			},
+			returnIs: assert.Equal,
+			returnVal: CacheSettings{
+				CacheMaxAge: "600s",
+				TimeRange: &backend.TimeRange{
+					From: time.Date(2019, 7, 30, 20, 0, 0, 0, time.UTC),
+					To:   time.Date(2019, 7, 30, 21, 10, 0, 0, time.UTC),
+				},
+			},
+		},
+		{
+			name: "should adjust settings based on query bin size 1h",
+			configuration: dataSourceData{
+				CacheMaxAge:    "10s",
+				DynamicCaching: true,
+			},
+			query: `SampleData\n` +
```

Comment:
I think the diff ends up funny when this fails with using `+`, so better to use three backticks instead